lookout/style/format/analyzer.py

```diff
@@ -64,23 +113,50 @@ def analyze(self, ptr_from: ReferencePointer, ptr_to: ReferencePointer,
                 fe = FeatureExtractor(language=lang, **rules.origin_config["feature_extractor"])
                 res = fe.extract_features([file], lines)
                 if res is None:
-                    comment = Comment()
-                    comment.file = file.path
-                    comment.confidence = 100
-                    comment.line = 1
-                    comment.text = "Failed to parse this file"
+                    if self.config["report_parse_failures"]:
+                        comment = Comment()
+                        comment.file = file.path
+                        comment.confidence = 100
+                        comment.line = 1
+                        comment.text = "Failed to parse this file"
+                        comment.append(comment)
+                    self.log.warning("Failed to parse %s", file.path)
                     continue
-                X, y, vnodes_y, _ = res
-                self.log.debug("predicting values for %d samples", len(y))
-                y_pred, winners = rules.predict(X, True)
+                X, y, vnodes_y, vnodes = res
+                X, _ = fe.select_features(X, y)
+                y_pred, rule_winners = rules.predict(X, True)
                 assert len(y) == len(y_pred)
 
-                for yi, y_predi, vnode, winner in zip(y, y_pred, vnodes_y, winners):
-                    if yi != y_predi:
-                        comment = vnode.to_comment(y_predi)
-                        comment.file = file.path
-                        comment.confidence = int(round(rules.rules[winner].stats.conf * 100))
-                        comments.append(comment)
+                code_lines = file.content.decode("utf-8", "replace").splitlines()
+                for line_nodes in group_line_nodes(y, y_pred, vnodes_y, rule_winners):
+                    code_line_number = line_nodes[0][2].start.line  # 1-based
+                    if self.config["report_triggered_rules"]:
+                        code_text = ""
+                        if self.config["report_code_lines"]:
+                            code_text = get_code_chunk(lang, code_lines, code_line_number)
+                        vnodes_comments = [
+                            "%s\n%s\n" % (get_error_description(vnode, y_predi),
```

Comment:
the `\n` is a joining character, it should probably be part of the join call rather than being here